### DISCUSSION TITLE
TileCreator.getTile gives a distorted image for a particular size

### DISCUSSION THREAD

**Discussion by RanjitaAcharya** [ngageoint/geopackage-java#55] [Q&A] [answered]
Hi,

I am using TileCreator.getTile method using boundary box value and particular zoom level. 
Also while creating TileCreator I am providing height and width .
For zoom level 5 , height = 8192 and width = 8192 the image is coming is distorted . If I increase or decrease the size the image is coming all fine . 
Projection used is : EPSG_WEB_MERCATOR
Is it a known issue or is there any restriction on the height and width for TileCreator ?

**Comment by bosborn** [comment]
Like a map server, the requested bounds provided are stretched to the desired width and height.  For Web Mercator, the width and height of the bounding box would need to be equal to create a non "distorted" / stretched image.

Example:
```
TileCreator tileCreator = new TileCreator(tileDao, 512, 512,
		ProjectionFactory.getProjection(ProjectionConstants.EPSG_WEB_MERCATOR),
		ImageUtils.IMAGE_FORMAT_PNG);

// Width: 1252344.0, Height: 1252344.0
GeoPackageTile tile = tileCreator.getTile(new BoundingBox(-10018754, 5009377, -8766410, 6261721), 5);
```
![tile](https://user-images.githubusercontent.com/11667761/146091569-8716e875-a82a-45b1-b27d-04b71d824d01.png)
```
// Width: 1252344.0, Height: 2504688.0
GeoPackageTile tile = tileCreator.getTile(new BoundingBox(-10018754, 3757033, -8766410, 6261721), 5);
```
![tile2](https://user-images.githubusercontent.com/11667761/146091600-8f7b0d03-259f-44eb-b510-9906e00d0b73.png)
```
// Width: 2504688.0, Height: 1252344.0
GeoPackageTile tile = tileCreator.getTile(new BoundingBox(-10018754, 5009377, -7514066, 6261721), 5);
```
![tile3](https://user-images.githubusercontent.com/11667761/146091620-eb9343e0-0501-463d-8374-341621f306ad.png)

A web server example with the same bounds:

https://services.arcgisonline.com/arcgis/rest/services/Elevation/World_Hillshade/MapServer/export?transparent=true&format=png&bbox=-10018754,5009377,-8766410,6261721&bboxSR=3857&imageSR=3857&size=512,512&f=image

https://services.arcgisonline.com/arcgis/rest/services/Elevation/World_Hillshade/MapServer/export?transparent=true&format=png&bbox=-10018754,3757033,-8766410,6261721&bboxSR=3857&imageSR=3857&size=512,512&f=image

https://services.arcgisonline.com/arcgis/rest/services/Elevation/World_Hillshade/MapServer/export?transparent=true&format=png&bbox=-10018754,5009377,-7514066,6261721&bboxSR=3857&imageSR=3857&size=512,512&f=image

If you want to use a non equal width and height, request a bounding box equal in ratio.

```
TileCreator tileCreator = new TileCreator(tileDao, 256, 512,
		ProjectionFactory.getProjection(ProjectionConstants.EPSG_WEB_MERCATOR),
		ImageUtils.IMAGE_FORMAT_PNG);

// Width: 1252344.0, Height: 2504688.0
GeoPackageTile tile = tileCreator.getTile(new BoundingBox(-10018754, 3757033, -8766410, 6261721), 5);
```
![tile4](https://user-images.githubusercontent.com/11667761/146092384-8c7c06c4-399a-458b-92a2-b6ce86c6adb4.png)

**Comment by bosborn** [comment, marked as answer]
After looking, I'm adding [some changes](https://github.com/ngageoint/geopackage-java/commit/c435a79dda06c1d4bcfe94c059d4f770e9e5c850) in the next release for improved handling when tile width and/or height **are not** specified.  When only one of tile width or height is provided, a determination of the other using the requested bounds aspect ratio.  When neither tile width or height are provided, an approximated pixel length based upon the saved tiles is adjusted to the requested bounds aspect ratio.  Currently before this change, the tile width and/or height from the tile matrix table is used when not provided.

Some examples for **version 6.1.2+~~, once released~~**.

No height
```
TileCreator tileCreator = new TileCreator(tileDao, 256, null,
				ProjectionFactory.getProjection(ProjectionConstants.EPSG_WEB_MERCATOR),
				ImageUtils.IMAGE_FORMAT_PNG);
GeoPackageTile tile = tileCreator.getTile(new BoundingBox(-10018754, 3757033, -8766410, 6261721), 5);
```
256 × 512
![tile5](https://user-images.githubusercontent.com/11667761/146266250-e7ffffe0-be31-414b-9a9d-b1ce40ee8f2e.png)

No width
```
TileCreator tileCreator = new TileCreator(tileDao, null, 256,
				ProjectionFactory.getProjection(ProjectionConstants.EPSG_WEB_MERCATOR),
				ImageUtils.IMAGE_FORMAT_PNG);
GeoPackageTile tile = tileCreator.getTile(new BoundingBox(-10018754, 3757033, -8766410, 6261721), 5);
```
128 × 256
![tile6](https://user-images.githubusercontent.com/11667761/146266466-698866d4-0986-430d-b90b-6712b4c36a9e.png)

No width or height, x=8, y=11-12
```
TileCreator tileCreator = new TileCreator(tileDao,
				ProjectionFactory.getProjection(ProjectionConstants.EPSG_WEB_MERCATOR),
				ImageUtils.IMAGE_FORMAT_PNG);
BoundingBox boundingBox = TileBoundingBoxUtils
				.getWebMercatorBoundingBox(new TileGrid(8, 11, 8, 12), 5);
GeoPackageTile tile = tileCreator.getTile(boundingBox, 5);
```
256 × 512
![tile10](https://user-images.githubusercontent.com/11667761/146266955-753a9531-e095-49c5-8846-c55e1cbc7bd1.png)

No width or height, WGS84, x=8, y=11-12
```
TileCreator tileCreator = new TileCreator(tileDao,
				ProjectionFactory.getProjection(ProjectionConstants.EPSG_WORLD_GEODETIC_SYSTEM),
				ImageUtils.IMAGE_FORMAT_PNG);
BoundingBox boundingBox = TileBoundingBoxUtils
				.getWebMercatorBoundingBox(new TileGrid(8, 11, 8, 12), 5);
boundingBox = boundingBox.transform(
				GeometryTransform.create(ProjectionConstants.EPSG_WEB_MERCATOR,
						ProjectionConstants.EPSG_WORLD_GEODETIC_SYSTEM));
GeoPackageTile tile = tileCreator.getTile(boundingBox, 5);
```
338 × 510
![tile7](https://user-images.githubusercontent.com/11667761/146267288-ae874e57-1f1a-462b-a75a-b7b014b85cc4.png)

No width or height, WGS84, WGS84 x=16, y= 7
```
TileCreator tileCreator = new TileCreator(tileDao,
				ProjectionFactory.getProjection(ProjectionConstants.EPSG_WORLD_GEODETIC_SYSTEM),
				ImageUtils.IMAGE_FORMAT_PNG);
GeoPackageTile tile = tileCreator.getTile(TileBoundingBoxUtils.getWGS84BoundingBox(16, 7, 5), 5);
```
190 × 190
[See comparison](https://osm.gs.mil/tiles/default_pc/5/16/7.png)
![tile8](https://user-images.githubusercontent.com/11667761/146267587-a17bb140-7e6b-43b5-aa97-f84eb41a4a05.png)

**Reply by bosborn** [reply]
These updates released in [6.1.2](https://github.com/ngageoint/geopackage-java/releases/tag/6.1.2)

**Reply by RanjitaAcharya** [reply]
Thanks for the update !
